File: src/js/classes/damageOverTime.js
// Damage over time auras
class DamageOverTime {
  constructor (player) {
    this.player = player
    this.durationTotal = 0 // Total duration of the dot
    this.tickTimerTotal = 3 // Total duration of each tick (default is 3 seconds between ticks)
    this.tickTimerRemaining = 0 // Time until next tick
    this.ticksRemaining = 0 // Amount of ticks remaining before the dot expires
    this.snapshots = true
    this.dmg = 0
    this.spellPower = 0 // Amount of Spell Power (plus Shadow or Fire Power) when the dot was applied
    this.modifier = 1
    this.active = false
    this.name = null
    this.coefficient = 0
    this.amplified = false // Amplify Curse
  }

  setup () {
    this.varName = camelCase(this.name)
    this.originalDurationTotal = this.durationTotal
    // T4 4pc
    if ((this.varName == 'immolate' || this.varName == 'corruption') && this.player.sets['645'] >= 4) {
      this.durationTotal += 3
    }
    this.ticksTotal = Math.round(this.durationTotal / this.tickTimerTotal)
    this.player.damageBreakdown[this.varName] = this.player.damageBreakdown[this.varName] || { name: this.name }
    this.player.auraBreakdown[this.varName] = this.player.auraBreakdown[this.varName] || { name: this.name }
  }

  apply () {
    if (this.active) {
      this.player.combatLog(this.name + ' refreshed before letting it expire')
    } else {
      // Keep a timestamp of when the aura was applied so we can calculate the uptime when it fades
      this.player.auraBreakdown[this.varName].appliedAt = this.player.fightTime
    }
    this.player.auraBreakdown[this.varName].count = this.player.auraBreakdown[this.varName].count + 1 || 1
    const refreshedOrApplied = this.active ? 'refreshed' : 'applied'
    this.active = true
    this.tickTimerRemaining = this.tickTimerTotal
    this.ticksRemaining = this.ticksTotal
    let spellPower = this.player.getSpellPower() + this.player.stats[this.school + 'Power']
    this.spellPower = spellPower
    this.player.combatLog(this.name + ' ' + refreshedOrApplied + ' (' + Math.round(spellPower) + ' Spell Power)')

    // Siphon Life snapshots the presence of ISB. So if ISB isn't up when it's cast, it doesn't get the benefit even if it comes up later during the duration.
    if (this.varName == 'siphonLife') {
      this.isbActive = (this.player.auras.improvedShadowBolt && this.player.auras.improvedShadowBolt.active && this.player.simSettings.customIsbUptime == 'no')
    }
    // Amplify Curse
    if ((this.varName == 'curseOfAgony' || this.varName == 'curseOfDoom') && this.player.auras.amplifyCurse && this.player.auras.amplifyCurse.active) {
      this.amplified = true
      this.player.auras.amplifyCurse.fade()
    } else {
      this.amplified = false
    }
  }

  fade (endOfIteration = false) {
    this.active = false
    this.tickTimerRemaining = 0
    this.ticksRemaining = 0
    const auraUptime = this.player.fightTime - this.player.auraBreakdown[this.varName].appliedAt
    this.player.auraBreakdown[this.varName].uptime = this.player.auraBreakdown[this.varName].uptime + auraUptime || auraUptime
    if (!endOfIteration) {
      this.player.combatLog(this.name + ' faded')
    }
  }

  getModifier () {
    let modifier = this.modifier * this.player.stats[this.school + 'Modifier']
    // Amplify Curse
    if (this.amplified) {
      modifier *= 1.5
    }
    // Improved Shadow Bolt
    if ((this.school == 'shadow' && this.player.auras.improvedShadowBolt && this.player.auras.improvedShadowBolt.active && this.varName != 'siphonLife') || (this.varName == 'siphonLife' && this.isbActive)) {
      modifier *= this.player.auras.improvedShadowBolt.modifier
    }
    return modifier
  }

  getConstantDamage() {
    const spellPower = this.snapshots ? this.spellPower : this.player.getSpellPower() + this.player.stats[this.school + 'Power']
    let modifier = this.getModifier()
    const partialResistMultiplier = this.player.getPartialResistMultiplier(this.player.enemy[this.school + 'Resist'])
    const dmg = (this.dmg + spellPower * this.coefficient) * modifier * partialResistMultiplier

    return [dmg, spellPower, modifier, partialResistMultiplier]
  }

  // Predicts how much damage the dot will do over its full duration
  predictDamage() {
    return this.getConstantDamage()[0]
  }

  tick (t) {
    if (this.active) {
      this.tickTimerRemaining = Math.max(0, this.tickTimerRemaining - t)

      if (this.tickTimerRemaining == 0) {
        const constantDamage = this.getConstantDamage()
        const dmg = constantDamage[0] / (this.originalDurationTotal / this.tickTimerTotal)
        const spellPower = constantDamage[1]
        const modifier = constantDamage[2]
        const partialResistMultiplier = constantDamage[3]

        // Check for Nightfall proc
        if (this.varName == 'corruption' && this.player.talents.nightfall > 0) {
          if (random(1, 100) <= this.player.talents.nightfall * 2) {
            this.player.auras.shadowTrance.apply()
          }
        }

        this.player.damageBreakdown[this.varName].damage = this.player.damageBreakdown[this.varName].damage + dmg || dmg
        this.player.iterationDamage += dmg
        this.player.combatLog(this.name + ' Tick ' + Math.round(dmg) + ' (' + this.dmg + ' Base Damage - ' + Math.round(spellPower) + ' Spell Power - ' + this.coefficient + ' Coefficient - ' + (Math.round(modifier * 10000) / 100).toFixed(2) + '% Damage Modifier ' + Math.round(partialResistMultiplier * 1000) / 1000 + '% Partial Resist Multiplier)')
        this.ticksRemaining--
        this.tickTimerRemaining = this.tickTimerTotal

        // Ashtongue Talisman of Shadows
        if (this.varName == 'corruption' && this.player.auras.ashtongueTalismanOfShadows && random(1, 100) <= this.player.auras.ashtongueTalismanOfShadows.procChance) {
          this.player.auras.ashtongueTalismanOfShadows.apply()
        }

        // Timbal's Focusing Crystal
        if (this.player.spells.timbalsFocusingCrystal && this.player.spells.timbalsFocusingCrystal.cooldownRemaining <= 0 && random(1, 100) <= this.player.spells.timbalsFocusingCrystal.procChance) {
          this.player.spells.timbalsFocusingCrystal.startCast()
        }

        if (this.ticksRemaining <= 0) {
          this.fade()
        }
      }
    }
  }
}

class CorruptionDot extends DamageOverTime {
  constructor (player) {
    super(player)
    this.durationTotal = 18
    this.tickTimerTotal = 3
    this.dmg = 900
    this.school = 'shadow'
    this.name = 'Corruption'
    this.coefficient = 0.936 + (0.12 * player.talents.empoweredCorruption)
    this.minimumDuration = 9
    this.setup()

    // T3 4pc
    if (player.sets['529'] >= 4) this.modifier *= 1.12
  }

  getModifier () {
    let modifier = super.getModifier()
    if (this.player.talents.shadowMastery > 0 && this.player.talents.contagion > 0) {
      // Divide away the bonus from Shadow Mastery
      modifier /= (1 + (this.player.talents.shadowMastery * 0.02))
      // Multiply the modifier with the bonus from Shadow Mastery + Contagion
      modifier *= (1 * (1 + ((this.player.talents.shadowMastery * 0.02) + (this.player.talents.contagion / 100))))
    }
    // Add the t5 4-set bonus modifier
    modifier *= this.t5BonusModifier
    return modifier
  }

  apply (spellPower) {
    // T5 4-set bonus modifier
    this.t5BonusModifier = 1
    super.apply(spellPower)
  }
}

class UnstableAfflictionDot extends DamageOverTime {
  constructor (player) {
    super(player)
    this.durationTotal = 18
    this.tickTimerTotal = 3
    this.dmg = 1050
    this.school = 'shadow'
    this.name = 'Unstable Affliction'
    this.coefficient = 1.2
    this.minimumDuration = 9
    this.setup()
  }
}

class SiphonLifeDot extends DamageOverTime {
  constructor (player) {
    super(player)
    this.durationTotal = 30
    this.tickTimerTotal = 3
    this.dmg = 630
    this.school = 'shadow'
    this.name = 'Siphon Life'
    this.coefficient = 1
    this.minimumDuration = 30
    this.setup()
  }
}

class ImmolateDot extends DamageOverTime {
  constructor (player) {
    super(player)
    this.durationTotal = 15
    this.tickTimerTotal = 3
    this.dmg = 615
    this.school = 'fire'
    this.name = 'Immolate'
    this.coefficient = 0.65
    this.minimumDuration = 12
    this.setup()
  }

  getModifier () {
    return super.getModifier() * this.t5BonusModifier
  }

  apply (spellPower) {
    // T5 4-set bonus modifier
    this.t5BonusModifier = 1
    super.apply(spellPower)
  }
}

class CurseOfAgonyDot extends DamageOverTime {
  constructor (player) {
    super(player)
    this.durationTotal = 24
    this.tickTimerTotal = 3
    this.dmg = 1356
    this.school = 'shadow'
    this.name = 'Curse of Agony'
    this.coefficient = 1.2
    this.minimumDuration = 15
    this.setup()
  }

  getModifier () {
    let modifier = super.getModifier()
    // Remove bonus from Shadow Mastery and add bonus from Shadow Mastery + Contagion + Improved Curse of Agony
    modifier /= (1 + (this.player.talents.shadowMastery * 0.02))
    modifier *= (1 * (1 + ((this.player.talents.shadowMastery * 0.02) + (this.player.talents.contagion / 100) + (this.player.talents.improvedCurseOfAgony * 0.05))))
    return modifier
  }
}

class CurseOfDoomDot extends DamageOverTime {
  constructor (player) {
    super(player)
    this.durationTotal = 60
    this.tickTimerTotal = 60
    this.dmg = 4200
    this.school = 'shadow'
    this.name = 'Curse of Doom'
    this.coefficient = 2
    this.setup()
  }
}

File: src/cpp/damageOverTime.cpp
#include "damageOverTime.h"
#include "common.h"
#include "player.h"

DamageOverTime::DamageOverTime(Player* player) : player(player)
{
    duration = 0;
    tickTimerTotal = 3;
    tickTimerRemaining = 0;
    ticksRemaining = 0;
    dmg = 0;
    spellPower = 0;
    modifier = 1;
    active = false;
    coefficient = 0;
    amplified = false;
}

void DamageOverTime::setup()
{
    originalDuration = duration;

    // T4 4pc
    if ((name == "Corruption" || name == "Immolate") && player->sets->t4 >= 4)
    {
        duration += 3;
    }

    ticksTotal = duration / tickTimerTotal;
    
    if (player->combatLogBreakdown.count(name) == 0)
    {
        player->combatLogBreakdown.insert(std::make_pair(name, new CombatLogBreakdown(name)));
    }
}

void DamageOverTime::apply()
{
    if (active && player->shouldWriteToCombatLog())
    {
        std::string msg = name + " refreshed before letting it expire";
    }
    else if (!active)
    {
        player->combatLogBreakdown.at(name)->appliedAt = player->fightTime;
    }
    bool isActive = active;
    int spellPower = player->getSpellPower(school);
    this->spellPower = spellPower;

    player->combatLogBreakdown.at(name)->count++;
    active = true;
    tickTimerRemaining = tickTimerTotal;
    ticksRemaining = ticksTotal;

    if (player->shouldWriteToCombatLog())
    {
        std::string refreshedOrApplied = isActive ? "refreshed" : "applied";
        std::string msg = name + " " + refreshedOrApplied + " (" + std::to_string(spellPower) + " Spell Power)";
        player->combatLog(msg);
    }
    // Siphon Life snapshots the presence of ISB. So if ISB isn't up when it's cast, it doesn't get the benefit even if it comes up later during the duration.
    if (player->spells->SiphonLife != NULL && name == player->spells->SiphonLife->name)
    {
        isbActive = !player->settings->usingCustomIsbUptime && player->auras->ImprovedShadowBolt != NULL && player->auras->ImprovedShadowBolt->active;
    }
    // Amplify Curse
    if (((player->spells->CurseOfAgony != NULL && name == player->spells->CurseOfAgony->name) || (player->spells->CurseOfDoom != NULL && name == player->spells->CurseOfDoom->name)) && player->auras->AmplifyCurse != NULL && player->auras->AmplifyCurse->active)
    {
        amplified = true;
        player->auras->AmplifyCurse->fade();
    }
    else
    {
        amplified = false;
    }
}

void DamageOverTime::fade(bool endOfIteration)
{
    active = false;
    tickTimerRemaining = 0;
    ticksRemaining = 0;
    
    double auraUptime = player->fightTime - player->combatLogBreakdown.at(name)->appliedAt;
    player->combatLogBreakdown.at(name)->uptime += auraUptime;

    if (!endOfIteration && player->shouldWriteToCombatLog())
    {
        std::string msg = name + " faded";
        player->combatLog(msg);
    }
}

double DamageOverTime::getModifier()
{
    double dmgModifier = modifier;
    if (school == SpellSchool::SHADOW)
    {
        dmgModifier *= player->stats->shadowModifier;
    }
    else if (school == SpellSchool::FIRE)
    {
        dmgModifier *= player->stats->fireModifier;
    }
    // Improved Shadow Bolt
    if ((school == SpellSchool::SHADOW && player->auras->ImprovedShadowBolt != NULL && player->auras->ImprovedShadowBolt->active && (player->spells->SiphonLife == NULL || name != player->spells->SiphonLife->name)) || (player->spells->SiphonLife != NULL && name == player->spells->SiphonLife->name && isbActive))
    {
        dmgModifier *= player->auras->ImprovedShadowBolt->modifier;
    }
    return dmgModifier;
}

std::vector<double> DamageOverTime::getConstantDamage()
{
    double spellPower = this->spellPower;
    // If the DoT isn't currently active then this.spellPower will be 0, so use the player's current Spell Power
    if (!active)
    {
        spellPower = player->getSpellPower(school);
    }
    double modifier = getModifier();
    double partialResistMultiplier = player->getPartialResistMultiplier(school);
    double baseDamage = dmg;
    // Amplify Curse
    if (amplified)
    {
        baseDamage *= 1.5;
    }
    // Add the t5 4pc bonus modifier to the base damage
    if (((player->spells->Corruption != NULL && name == player->spells->Corruption->name) || (player->spells->Immolate != NULL && name == player->spells->Immolate->name)) && player->sets->t5 >= 4)
    {
        baseDamage *= t5BonusModifier;
    }
    double totalDamage = baseDamage;
    totalDamage += spellPower * coefficient;
    totalDamage *= modifier * partialResistMultiplier;

    return std::vector<double> { baseDamage, totalDamage, spellPower, modifier, partialResistMultiplier };
}

// Predicts how much damage the dot will do over its full duration
double DamageOverTime::predictDamage()
{
    std::vector<double> constantDamage = getConstantDamage();
    double dmg = constantDamage[1];
    // If it's Corruption or Immolate then divide by the original duration (18s and 15s) and multiply by the durationTotal property
    // This is just for the t4 4pc bonus since their durationTotal property is increased by 3 seconds to include another tick
    // but the damage they do stays the same which assumes the normal duration without the bonus
    if ((player->spells->Corruption != NULL && name == player->spells->Corruption->name) || (player->spells->Immolate != NULL && name == player->spells->Immolate->name))
    {
        dmg /= originalDuration;
        dmg *= duration;
    }
    
    return dmg;
}

void DamageOverTime::tick(double t)
{
    tickTimerRemaining -= t;

    if (tickTimerRemaining <= 0)
    {
        std::vector<double> constantDamage = getConstantDamage();
        double baseDamage = constantDamage[0];
        double dmg = constantDamage[1] / (originalDuration / tickTimerTotal);
        double spellPower = constantDamage[2];
        double modifier = constantDamage[3];
        double partialResistMultiplier = constantDamage[4];

        // Check for Nightfall proc
        if (player->spells->Corruption != NULL && name == player->spells->Corruption->name && player->talents->nightfall > 0)
        {
            if (player->getRand() <= player->talents->nightfall * 2 * player->critChanceMultiplier)
            {
                player->auras->ShadowTrance->apply();
            }
        }

        player->addIterationDamageAndMana(name, 0, dmg);
        player->iterationDamage += dmg;
        ticksRemaining--;
        tickTimerRemaining = tickTimerTotal;

        if (player->shouldWriteToCombatLog())
        {
            std::string msg = name + " Tick " + truncateTrailingZeros(std::to_string(round(dmg))) + " (" + truncateTrailingZeros(std::to_string(baseDamage)) + " Base Damage - " + truncateTrailingZeros(std::to_string(spellPower)) + " Spell Power - " + truncateTrailingZeros(std::to_string(coefficient), 3) + " Coefficient - " + truncateTrailingZeros(std::to_string(round(modifier * 10000) / 100), 3) + "% Damage Modifier - " + truncateTrailingZeros(std::to_string(round(partialResistMultiplier * 1000) / 10)) + "% Partial Resist Multiplier";
            if (t5BonusModifier > 1)
            {
                msg += " - " + std::to_string(round(t5BonusModifier * 10000) / 100) + "% Base Dmg Modifier (T5 4pc bonus)";
            }
            msg += ")";
            player->combatLog(msg);
        }

        // Ashtongue Talisman of Shadows
        if (player->spells->Corruption != NULL && name == player->spells->Corruption->name && player->auras->AshtongueTalismanOfShadows != NULL && player->getRand() <= player->auras->AshtongueTalismanOfShadows->procChance * player->critChanceMultiplier)
        {
            player->auras->AshtongueTalismanOfShadows->apply();
        }
        // Timbal's Focusing Crystal
        if (player->spells->TimbalsFocusingCrystal != NULL && player->spells->TimbalsFocusingCrystal->cooldownRemaining <= 0 && player->getRand() <= player->spells->TimbalsFocusingCrystal->procChance * player->critChanceMultiplier)
        {
            player->spells->TimbalsFocusingCrystal->startCast();
        }

        if (ticksRemaining <= 0)
        {
            fade();
        }
    }
}

CorruptionDot::CorruptionDot(Player* player) : DamageOverTime(player)
{
    name = "Corruption";
    duration = 18;
    tickTimerTotal = 3;
    dmg = 900;
    school = SpellSchool::SHADOW;
    coefficient = 0.936 + (0.12 * player->talents->empoweredCorruption);
    minimumDuration = 9;
    t5BonusModifier = 1;
    setup();

    // T3 4pc
    if (player->sets->plagueheart >= 4)
    {
        modifier *= 1.12;
    }
}

double CorruptionDot::getModifier()
{
    double modifier = DamageOverTime::getModifier();
    if (player->talents->shadowMastery > 0 && player->talents->contagion > 0)
    {
        // Divide away the bonus from Shadow Mastery
        modifier /= (1 + (player->talents->shadowMastery * 0.02));
        // Multiply the modifier with the bonus from Shadow Mastery + Contagion
        modifier *= (1 * (1 + ((player->talents->shadowMastery * 0.02) + (player->talents->contagion / 100.0))));
    }
    return modifier;
}

void CorruptionDot::apply()
{
    // T5 4pc bonus modifier
    t5BonusModifier = 1;
    DamageOverTime::apply();
}

UnstableAfflictionDot::UnstableAfflictionDot(Player* player) : DamageOverTime(player)
{
    name = "Unstable Affliction";
    duration = 18;
    tickTimerTotal = 3;
    dmg = 1050;
    school = SpellSchool::SHADOW;
    coefficient = 1.2;
    minimumDuration = 9;
    setup();
}

SiphonLifeDot::SiphonLifeDot(Player* player) : DamageOverTime(player)
{
    name = "Siphon Life";
    duration = 30;
    tickTimerTotal = 3;
    dmg = 630;
    school = SpellSchool::SHADOW;
    coefficient = 1;
    minimumDuration = 30;
    setup();
}

ImmolateDot::ImmolateDot(Player* player) : DamageOverTime(player)
{
    name = "Immolate";
    duration = 15;
    tickTimerTotal = 3;
    dmg = 615;
    school = SpellSchool::FIRE;
    coefficient = 0.65;
    minimumDuration = 12;
    t5BonusModifier = 1;
    setup();
}

void ImmolateDot::apply()
{
    // T5 4pc bonus modifier
    t5BonusModifier = 1;
    DamageOverTime::apply();
}

CurseOfAgonyDot::CurseOfAgonyDot(Player* player) : DamageOverTime(player)
{
    name = "Curse of Agony";
    duration = 24;
    tickTimerTotal = 3;
    dmg = 1356;
    school = SpellSchool::SHADOW;
    coefficient = 1.2;
    minimumDuration = 15;
    setup();
}

double CurseOfAgonyDot::getModifier()
{
    double modifier = DamageOverTime::getModifier();
    // Remove bonus from Shadow Mastery and add bonus from Shadow Mastery + Contagion + Improved Curse of Agony
    modifier /= (1 + (player->talents->shadowMastery * 0.02));
    modifier *= (1 * (1 + ((player->talents->shadowMastery * 0.02) + (player->talents->contagion / 100.0) + (player->talents->improvedCurseOfAgony * 0.05))));
    return modifier;
}

CurseOfDoomDot::CurseOfDoomDot(Player* player) : DamageOverTime(player)
{
    name = "Curse of Doom";
    duration = 60;
    tickTimerTotal = 60;
    dmg = 4200;
    school = SpellSchool::SHADOW;
    coefficient = 2;
    minimumDuration = 60;
    setup();
}

double CurseOfDoomDot::getModifier()
{
    double modifier = DamageOverTime::getModifier();

    // CoD doesn't benefit from SM
    if (player->talents->shadowMastery > 0)
    {
        modifier /= (1 + (0.02 * player->talents->shadowMastery));
    }

    return modifier;
}